californium-core/src/main/java/org/eclipse/californium/core/network/stack/TCPBlockwiseLayer.java

```diff
@@ -0,0 +1,394 @@
+package org.eclipse.californium.core.network.stack;
+
+import java.nio.ByteBuffer;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+import org.eclipse.californium.core.coap.BlockOption;
+import org.eclipse.californium.core.coap.OptionSet;
+import org.eclipse.californium.core.coap.Request;
+import org.eclipse.californium.core.coap.Response;
+import org.eclipse.californium.core.coap.CoAP.ResponseCode;
+import org.eclipse.californium.core.network.Exchange;
+import org.eclipse.californium.core.network.config.NetworkConfig;
+import org.eclipse.californium.core.network.config.NetworkConfig.Keys;
+
+/**
+ * Blockwise layer for TCP. Extended mainly for BERT option. Capable of handling
+ * incoming blocks with SZX = 7.
+ * 
+ * @author Pratheek Rai.
+ *
+ */
+public class TCPBlockwiseLayer extends BlockwiseLayer {
+
+	/**
+	 * logger.
+	 */
+	private static final Logger LOGGER = Logger.getLogger(TCPBlockwiseLayer.class.getName());
+
+	/**
+	 * BERT SZX constant.
+	 */
+	private final static int BERT_SZX = 7;
+	
+	/**
+	 * Internal block size for BERT. i.e internally BERT is handled similar to
+	 * SZX = 6.
+	 */
+	private final static int BERT_INT_BLOCK_SIZE = 1024;
+
+	/**
+	 * The number of 1024 Bytes packets to be sent as a block. Size of 1 BERT
+	 * block = bertStepSize * 1024 Bytes
+	 */
+	private int bertStepSize;
+
+	/**
+	 * Flag for BERT option.
+	 */
+	private boolean isBertEnabled;
+
+	/**
+	 * Constructor 
+	 * @param config - network configuration.
+	 */
+	public TCPBlockwiseLayer(NetworkConfig config) {
+		super(config);
+		bertStepSize = config.getInt(Keys.BERT_OPTION_STEP_SIZE, 1);
+		isBertEnabled = config.getBoolean(Keys.BERT_OPTION);
+		if (isBertEnabled) {
+			// Internally, BERT works works similar to SZX = 6.
+			preferredBlockSzx = 6;
+			preferredBlockSize = BERT_INT_BLOCK_SIZE;
+		}
+	}
+
+	/**
+	 * Handles inbound blockwise upload from peer. Handling is done here only
+	 * for SZX = 7 all other handling is done in the BlockwikseLayer.
+	 */
+	protected void handleInboundBlockwiseUpload(final Exchange exchange, final Request request) {
+
```

Comment:
This seems to be a lot of code copy from `BlockwiseLayer`. 
But as I meant previously, if we don't want to change the `BlockwiseLayer` itself too much, I agree that such a copy must then be used. May be we can get feedback from other committers, how much changes would be acceptable to omit code copies.